### ISSUE TITLE
Feature: Read someone's followers

### ISSUE THREAD

**Issue by DzianisH** [instabot-py/instabot.py#2052]
It will be useful to follow users that follows accounts like you. it would be nice to read followers (with pagination and etc.) of someones account

**Comment by mactrash** [comment]
+1 Good idea

**Comment by BartekChojnacki** [comment]
+1

**Comment by danielkiedrowski** [comment]
I remember that this was possible in an earlier version. Has this been thrown out?

**Comment by romarioagros** [comment]
+1. Please

**Comment by AlexanderProd** [comment]
@danielkiedrowski are you sure it was possible because I couldn't find anything about it in the codebase? I'd like to implement this feature I do have some questions though.

**Comment by marykrshaa** [comment]
> @danielkiedrowski are you sure it was possible because I couldn't find anything about it in the codebase? I'd like to implement this feature I do have some questions though.

yeah, i can vouch for him. i dont know why they decided to throw that out

**Comment by AlexanderProd** [comment]
@marykrshaa do you know in which Version it was?
In my experience with other bots this feature has the best performance since you can really fine tune your target audience, better than just with tags.

**Comment by AlexanderProd** [comment]
Update: this has been removed because the feature initially used an external service which stopped working when the instagram api changed (ink361).

**Comment by danielkiedrowski** [comment]
Yeah it was before the this bot used the api and remote controlled browser windows. Is this feature not available in the Instagram api?

**Comment by AlexanderProd** [comment]
@danielkiedrowski It seems like the problem is you'd have to gather a bunch of followers at once because you can't keep track of the position you are in someone else's followers, this on the other hand would probably cause you to get rate limited if done too fast. 

But I guess it would be possible to implement a mode where the bot goes through someone else's followers in realtime and follows them with a random sleep after each one. 

I'm new to the instagram api though and therefore I'm not sure if it works the way I expect it to.

**Comment by DzianisH** [comment]
Here is an example btw
https://github.com/instagrambot/instabot/blob/master/instabot/api/api.py#L478

**Comment by ZdravkoDimov** [comment]
Can you please help me! PLEASE! https://github.com/instabot-py/instabot.py/issues/2045

**Comment by AlexanderProd** [comment]
@DzianisH The difference is the repository you've linked uses the mobile API, this one uses the web API. It seems like the web API doesn't allow querying of all followers. This repository would need to be updated to use the mobile API.

**Comment by AlexanderProd** [comment]
> @DzianisH The difference is the repository you've linked uses the mobile API, this one uses the web API. It seems like the web API doesn't allow querying of all followers. This repository would need to be updated to use the mobile API.

Update: 
it seems like it is indeed possible since this web api wrapper supports a get users followers feature. 
https://github.com/OlegYurchik/pyInstagram/blob/9f30725fda7c10f523660fb119c7817a01d684e7/instagram/agents.py#L962

**Comment by Kravinskiy** [comment]
Ohh! This one could be really handy

**Comment by AKwoKWH** [comment]
Nice idea to follow someone's followers.
It is much more easier to target related audiences.

Also,
If it is possible to save a list of our own followers too?
Then the bot can compare the list and auto send a welcome message to new followers.

**Comment by dimisdas** [comment]
You cant send DMs over the web APl. For scraping and downloading user data in bulk there are better solutions like lnstaloader. There should also be arguements on how specific the scope of the bot library should be. It is _not_ a general APl wrapper, and if people think one should be necessary, perhaps a separate project should be created and loaded on instabot as a dependency.

**Comment by AKwoKWH** [comment]
Yes. 
It is better to create a "all in one bot" which is more user friendly.

I know there are other projects which use private API that can send DM, or even mute feed. 

However, it is quite troublesome to call another set of code to do these functions.
Say, I get the followers list from instabot_py then send msg using instapy.
Everything got duplicated.
(For example, we will hv 2 set of session files for the same account)

Loaded those features as from another projects as dependency should work, ( as the code is sort of "merged" into the original ).

**Comment by dimisdas** [comment]
There is no way to send a DM via InstaPy or
any bot that uses the web API. And unless there is a use case about a feature that fits the bot project, I’d argue against its implementation. 

This is not one size fits all project. This is a bot that emulates interactions. You wouldn’t manually make hundreds of requests to scrape data (which would get you banned).

Also, DMing people is an extremely bad practice that will get you banned, and will attract Instagram’s attention to this project. In my opinion, it should never be added on Instabot, even if it’s supported by the web API in the future.

**Comment by cala095** [comment]
I agree with dimisdas that DMing people is good way to get blocked, but the feature of following another user's followers it's really good, and it's more natural than following random users from hashtags, also every famous bots (stim, viki, etc.) base their algorithm on copying similar but bigger users fan bases.
I should end uni for summer, maybe I can help on developing something.

**Comment by ekkyu** [comment]
I developed code to get a follower and use it for targeting. 
This code uses the username to get the follower and target by getting the media_id of the follower.
I have confirmed that it works locally, but I think it is not perfect yet. 
As I send PR, please fix it!

**Comment by mediabots** [comment]
I have already built a python GUI bot about this feature
It can scrape followers/following of a target user id. 
At the end it would return a list of usernames.
Then one can follow those usernames

Scrapping other users followings or followers could be done by below code:
[NOTICE- this code would work for both Mobile & Web browser user-agents]

`
def get_follow_list_of_a_username(self,target_user_id,_type):

		if _type == "followers":
			private_hash = "c76146de99bb02f6415203be841dd25a" #for followers
			edge__by = "edge_followed_by"
		else:
			private_hash = "d04b0a864b4b54837c0d870b0e77e076" #for followings
			edge__by = "edge_follow"
		
		follow_list = []
		
		page_num = 1
		has_next_page = True
		after = ""
		first = 24
		print(_type) 
		
		while has_next_page:
			print(">>> Scanning {}: {} to {}".format(_type,((page_num-1)*first)+1,(page_num*first-1)+1))
			if not after:
				params = {'query_hash':private_hash,'variables':'{"include_highlight_reels":false,"include_reel":true,"id":'+target_user_id+',"fetch_mutual":true,"first":'+str(first)+'}'}
			else:
				params = {'query_hash':private_hash,'variables':'{"include_highlight_reels":false,"include_reel":true,"id":'+target_user_id+',"fetch_mutual":true,"first":'+str(first)+',"after":"'+after+'"}'}
			try
				resp,excep = session_func(self.session_main,url_graphql,params=params,redirects=False,proxies=proxies)
				if not excep:
					if resp and resp.status_code == 200:
						data = resp.json()
						has_next_page = bool(data["data"]["user"][edge__by]["page_info"]["has_next_page"])
						if has_next_page:
							after = data["data"]["user"][edge__by]["page_info"]["end_cursor"]
						user_lists = data["data"]["user"][edge__by]["edges"]	
						for user in user_lists:
							id_to_user[user["node"]["id"]] = user["node"]["username"]
							follow_list.append(user["node"]["username"])
						page_num+=1
					else:
						print("[Err] {} Page not responding properly! \r\nStatus Code: {}".format(resp.url,resp.status_code))
				else:
					print("[Exception] Exception! <during : get_follow_list_of_a_username()>")
			except Exception as err:
				print("[Exception] exception error <during : get_follow_list_of_a_username() >>> {}".format(err))
			time.sleep(random.randrange(5,10))
		# End of while loop
		
		return follow_list
`
`
def session_func(s,url,headers={},redirects=True,params=None,data=None,proxies=None,method="get",wait=random.randrange(5,10)):

	session_passed = True
	response = None
	if not s:
		s = requests.Session()
		session_passed =  False
	get = False
	tried = 1
	while not get and tried <= 3:
		#status_code = 200
		exception = False
		try:
			write_me_log ("[{}] Visiting {}".format(tried,url))
			if method == "get":
				if headers: # for API headers need to be different
					response = s.get(url,headers=headers,allow_redirects=redirects,params=params,verify=False, timeout=timeout,proxies=proxies)
				else:
					response = s.get(url,allow_redirects=redirects,params=params,verify=False, timeout=timeout,proxies=proxies)
			elif method == "post":
				response = s.post(url,verify=False, timeout=timeout,data=data,proxies=proxies)
			if response.status_code == 200:
				get = True
			else:
				time.sleep(30*tried) # or 30*tried
		except requests.ConnectionError as err:
			write_me_log("Your Internet Connection has some issue >>> {}".format(err))
			time.sleep(30*tried)
			exception = True
		except Exception as err:
			write_me_log("Exception occurred in session_func >>> {}".format(err))
			time.sleep(30*tried)
			exception = True
		tried+=1
	if not session_passed:
		s.close()
	#write_me_log("Wait in session_func")
	time.sleep(wait) #wait
	return response,exception
#End of the Function session_func()

`

You can find more about this on my Github repository : [https://github.com/mediabots/InstagramBot-GUI-Python](url)

**Comment by u0xpsec** [comment]
> @danielkiedrowski It seems like the problem is you'd have to gather a bunch of followers at once because you can't keep track of the position you are in someone else's followers, this on the other hand would probably cause you to get rate limited if done too fast.
> 
> But I guess it would be possible to implement a mode where the bot goes through someone else's followers in realtime and follows them with a random sleep after each one.
> 
> I'm new to the instagram api though and therefore I'm not sure if it works the way I expect it to.

could it pick usernames from a list provided by the user or from a username picked from the hashtag and randomly follow n numbers of their followers using the same logic of not following people already being followed?

**Comment by mifi** [comment]
Have been doing this for a few years with my library which also uses web automation https://github.com/mifi/instauto/blob/master/index.js#L306
